src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java

```diff
@@ -14,30 +14,166 @@
+  }
+
+  @VisibleForTesting
+  /** Creates and returns {@code lectureEntity} using parameters found in {@code request}. */
+  // TODO: Error check
+  protected Entity createLectureEntity(HttpServletRequest request) {
+    Optional<String> optionalLectureName = getParameter(request, NAME_INPUT);
+    String lectureName = optionalLectureName.isPresent() ? optionalLectureName.get() : "";
+    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
+    String videoUrl = optionalVideoUrl.isPresent() ? optionalVideoUrl.get() : "";
+    Optional<String> optionalVideoId = getVideoId(videoUrl);
+    String videoId = optionalVideoId.isPresent() ? optionalVideoId.get() : "";
+
+    Entity lectureEntity = new Entity(LectureEntityFields.KIND);
+    lectureEntity.setProperty(LectureEntityFields.NAME, lectureName);
+    lectureEntity.setProperty(LectureEntityFields.VIDEO_URL, videoUrl);
+    lectureEntity.setProperty(LectureEntityFields.VIDEO_ID, videoId);
+    return lectureEntity;
+  }
+
+  /** Returns lectures stored in the database. */
+  private List<Lecture> getLectures() {
+    Query query = new Query(LectureEntityFields.KIND);
+    PreparedQuery results = datastore.prepare(query);
+    List<Lecture> lectures = new ArrayList<>();
+    for (Entity lecture : results.asIterable()) {
+      Lecture newLecture = Lecture.fromLectureEntity(lecture);
+      lectures.add(newLecture);
+    }
+    return lectures;
+  }
+
+  /**
+   * Returns the value of {@code name} from the {@code request} form.
+   * If the {@code name} cannot be found, return {@code Optional.empty()}.
+   */
+  private Optional<String> getParameter(HttpServletRequest request, String name) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return Optional.empty();
+    }
+    return Optional.of(value);
+  }
+
+  @VisibleForTesting
+  /** Returns YouTube video ID for a given {@code videoUrl}. */
+  protected Optional<String> getVideoId(String videoUrl) {
+    Matcher matcher = videoUrlGeneratedPattern.matcher(videoUrl);
+    if (matcher.find()) {
+      return Optional.of(matcher.group());
+    }
+    // TODO: Throw an error saying ID not found.
+    return Optional.empty();
+  }
+
+  /**
+   * Returns URL redirecting to lecture view page with parameters {@code lectureId}
+   * and {@code videoId} found in {@code lectureEntity}.
+   */
+  private String buildRedirectUrl(Entity lectureEntity) {
+    String lectureId = String.valueOf(lectureEntity.getKey().getId());
+    String videoId = (String) lectureEntity.getProperty(LectureEntityFields.VIDEO_ID);
+
+    try {
+      URIBuilder urlBuilder = new URIBuilder(REDIRECT_URL)
+                                  .addParameter(LectureEntityFields.ID, lectureId)
+                                  .addParameter(LectureEntityFields.VIDEO_ID, videoId);
+      return urlBuilder.build().toString();
+    } catch (URISyntaxException urlBuilderError) {
+      throw new RuntimeException(urlBuilderError);
```

Comment:
We should avoid catching and rethrowing here. Also throwing a RuntimeException is generally discouraged. Can we return an optional possible here? So we force the caller to check if its present or not.

Reply:
fixed